New microservices image with working RunServices

TiagoTAlves opened the pull request:
This Pull Request creates a RunNumber, RunRegistryMeta and RunRegistryConfigs table which allows for the storage of run services data related to nanorc in daq-kube. 
These changes will be used to create a unified microservices image which will find the run services postgresql pod which has a database already deployed and add the tables to the already created database. These tables are created using flask_sqlalchemy.
All mentions of Oracle have been replaced with PostgresSQL.

The structure of these is as follows (for both both runnumber-rest and runregistry-rest):
* A rest.py - This calls upon the other python files and creates the applications
* A datbase.py - This creates the tables as python classes using flask_sqlalchemy
* A authentication.py - This creates a username and password to access the API's
* An api.py - This creates the api queries for accessing and inserting data into the database using sqlalchemy directly
* An entrypoint.sh - This  creates and entrypoint for the respective runservice in the image
* A deployment.yaml - This is a copy of the deployment.yaml file in daq-kube


This PR also requires the  [https://github.com/DUNE-DAQ/daq-kube/pull/35](url) to be approved too. 
 
TODO: Need to create new microservices dependencies image as well as microservices image

TiagoTAlves committed:
Database creation for daq-kube Microservices image

TiagoTAlves committed:
added db queries framework but commented out for now

TiagoTAlves committed:
corrected return out of function

TiagoTAlves committed:
making some changes to dockerfile so image can build, will change back afterwards

TiagoTAlves committed:
change from ssh back to https

TiagoTAlves committed:
continued changes to runserv microservices image

TiagoTAlves committed:
getting errors with oracle backend, testing with just postgres for now

TiagoTAlves committed:
removing oracle mentions

TiagoTAlves committed:
added sqlalchemy to requirements

TiagoTAlves committed:
added a backend fpr runnumber-rest

TiagoTAlves committed:
typo fixes

TiagoTAlves committed:
changes to rest.py and credentials.py

TiagoTAlves committed:
added sqlalchemy_utils as another required package

TiagoTAlves committed:
changing the postgresql port

TiagoTAlves committed:
indentation error

TiagoTAlves committed:
fixing microservices image

TiagoTAlves committed:
added MetaData to import list

TiagoTAlves committed:
more connection tests

TiagoTAlves committed:
more testing

TiagoTAlves committed:
add as a service

TiagoTAlves committed:
try a different service

TiagoTAlves committed:
remove engine from inside metadata

TiagoTAlves committed:
timezone errors

TiagoTAlves committed:
more timezone errors

TiagoTAlves committed:
psycopg2 added to engine

TiagoTAlves committed:
couldn't resolve host

TiagoTAlves committed:
debugging

jcpunk committed:
Attempt a code cleanup org for runnumber

jcpunk committed:
Try to facter out into parts

TiagoTAlves committed:
cleaned up runregistry in the same way as runnumber

TiagoTAlves committed:
changed scalar_one() to scalar() and removed credentials.py

TiagoTAlves committed:
chanegd run_number to rn so that it is more inline with RunNumber table

TiagoTAlves committed:
adding Column to import

TiagoTAlves committed:
made changes to requirements so all modules are compatible when creating dependency image

TiagoTAlves committed:
still unfinished chanegs to runregistry/api

TiagoTAlves committed:
Column is still complaining so I added the sqlalchemy module too

TiagoTAlves committed:
solving database errors

TiagoTAlves committed:
replaced sqlalchemy with flask_sqlalchemy modules

TiagoTAlves committed:
changed sqlite to postgresql

TiagoTAlves committed:
continued image problem solving

TiagoTAlves committed:
taking it back to what it was before

jcpunk committed:
Try to register db object with application

jcpunk committed:
Fix path error

TiagoTAlves committed:
reorganised db sqlalchemy

TiagoTAlves committed:
fix path error for runregistry

TiagoTAlves committed:
added SQLAlchemy as app

TiagoTAlves committed:
changing path to database

TiagoTAlves committed:
changing connection string and fixed circular db

TiagoTAlves committed:
change entrypoint for runregistry

jcpunk committed:
Drop Oracle LD libpath

jcpunk committed:
Drop obsolete comment

jcpunk committed:
Drop old interface files

jcpunk committed:
Drop old comments

jcpunk committed:
Minor nits in entrypoint.sh

jcpunk committed:
Don't enclose, just run unconditionally

jcpunk committed:
Reset name back to test mode

jcpunk committed:
Fix duplicate declare

TiagoTAlves committed:
changed deployments to match ones in daq-kube

jcpunk committed:
Minify and restore image spec

jcpunk committed:
Note required bits

TiagoTAlves committed:
added run number as a foreign key connected to RunNumber table

TiagoTAlves committed:
added a row for the inital run so that all subsequent runs can platform off of this one

TiagoTAlves committed:
changed some queries so its more similar to runnumber queries#

TiagoTAlves committed:
bringing dockerfile back to original in prep for PR

jcpunk committed:
Try to cleanup queries

TiagoTAlves committed:
Merge remote-tracking branch 'origin' into riehecky/runnumber_reorg

TiagoTAlves committed:
revert to working point

TiagoTAlves committed:
if no data, add a value starting at run 1

TiagoTAlves committed:
circular arguments makeing image not work

TiagoTAlves committed:
foreign key problem

jcpunk commented on runnumber-rest/api.py:
```diff
@@ -0,0 +1,110 @@
+@api.resource("/runnumber/get")
+class getRunNumber(Resource):
+    """
+    should return the run number in format: XXXXXXX
```
Can you fill out these comments in more detail?

TiagoTAlves replied:
Added in new commits, will keep this open in case further changes need to be made

jcpunk commented on runnumber-rest/api.py:
```diff
@@ -0,0 +1,110 @@
+from database import RunNumber
+from authentication import auth
+
+@api.resource("/runnumber/get")
```
The run registry has snippets that can be used to test, can you add those here?

TiagoTAlves replied:
Added in next commit

jcpunk commented on runregistry-rest/api.py:
```diff
@@ -0,0 +1,213 @@
+class getRunMeta(Resource):
+    """
+    Get XYZ
+    should return the RunMeta in format: XXXXXX
```
Can you fill out the comments of this style?

TiagoTAlves replied:
How do I upload a configuration to see what format it comes in?

jcpunk replied:
I think with a curl POST maybe?

TiagoTAlves replied:
I'll see what I can do then

jcpunk commented on runregistry-rest/rest.py:
```diff
@@ -1,237 +1,20 @@
-# As a testserver.
-# app.run(host= '0.0.0.0', port=5000, debug=True)
-# Normally spawned by gunicorn
+    app.run(host="0.0.0.0", port=5005, debug=DEBUG)
```
minor nit on missing new line at the end.

TiagoTAlves replied:
added

TiagoTAlves committed:
changes in response to PR comments

TiagoTAlves committed:
better comments for api query in run number

jcpunk commented on runnumber-rest/api.py:
```diff
@@ -0,0 +1,115 @@
+                db.select(func.max(RunNumber.rn))
+            ).scalar_one()
+            if current_max_run is None:
+                current_max_run = 1000
```
The `1000` here should probably be controlled by an environment variable in a similar way to how `DEBUG` is checked in the application startup code.

TiagoTAlves replied:
But if its controlled by an environment variable then surely that defeats the purpose of the run number being set automatically?

TiagoTAlves replied:
In the previous version of runservices we just artificially introduced a fake run so all the other runs could iterate up from it

TiagoTAlves replied:
Happy to change it either way though

jcpunk replied:
I was thinking something along the lines of:

```
                current_max_run = int(os.getenv("RUN_START", "1000"))
```

TiagoTAlves replied:
Added it in latest commit

TiagoTAlves committed:
final comments and commits changing the curl commands and adding more detail to comments

TiagoTAlves committed:
set initial runnumber as environment variable

jcpunk commented on runnumber-rest/api.py:
```diff
@@ -0,0 +1,115 @@
+            # the primary key sequence may not match
+            current_max_run = db.session.execute(
+                db.select(func.max(RunNumber.rn))
+            ).scalar_one()
```
This should probably be set to `scalar` since it can return `None`.

TiagoTAlves replied:
done

jcpunk committed:
black format

jcpunk commented on runnumber-rest/rest.py:
```diff
@@ -1,77 +1,20 @@
-        return resp
+To control set the following environment variables:
+DATABASE_URI - URI for sqlalchemy to use
+DEBUG - set to TRUE for flask debugging
```
Can you add the run start counter ENV to the note here?

TiagoTAlves replied:
done

jcpunk replied:
I'm not seeing this in the latest commit....

jcpunk commented on runregistry-rest/api.py:
```diff
@@ -0,0 +1,230 @@
+            return flask.make_response(str(e), 400)
+
+
+# $ -u fooUsr:barPass -X GET np04-srv-021:30015/runregistry/updatestop/<int:runNum>
```
Lost the `curl` from here...

jcpunk commented on runregistry-rest/api.py:
```diff
@@ -0,0 +1,230 @@
+class insertRun(Resource):
+    """
+    adds a run with the specified data given in the curl command
+    should return the XYZ in format: [1000, "foo", "bar", "dunedaq-vX.Y.Z", "@sspconf.tar.gz"]
```
If you could change XYZ to a more descriptive value, that would be great!

TiagoTAlves committed:
added missing curl, env variable description and changed scalar_one to scalar

jcpunk committed:
Simplify imports

jcpunk commented on runregistry-rest/database.py:
```diff
@@ -0,0 +1,27 @@
+
+
+class RunRegistryConfig(db.Model):
+    rn = db.Column(
```
Should this have a foreign key relationship to `RunRegistryMeta`?

TiagoTAlves replied:
Should there also be a foreignkey relationship with RunNumber database too?

jcpunk replied:
I'm not sure how to define foreign keys across applications, but I'd love for there to be some sort of sane relationship between them.

jcpunk replied:
Would it make sense to merge these two API's together?  Without `RunNumber` defined I'm not sure that foreign key would work as expected...

TiagoTAlves replied:
I may consider it, if I don't manage to figure this out

TiagoTAlves replied:
Looks like it didn't work as expected, should i actually merge the API's? Should I leave the foreignkeys to RunNumber seperate? Or should I present at the next CCM meeting and wait until we discuss? (I realise this may be a while before we have one of those meetings)

jcpunk replied:
I'm inclined towards merging, but we should probably discuss at CCM.

jcpunk commented on runregistry-rest/rest.py:
```diff
@@ -1,237 +1,21 @@
+To control set the following environment variables:
+DATABASE_URI - URI for sqlalchemy to use
+DEBUG - set to TRUE for flask debugging
+RUN_START - set to your starting run number, does not need to be changed after first launching as run number will be iterated from this initial value
```
I'm not showing `RUN_START` used by any of the code in runregistry....

TiagoTAlves replied:
I thought that it would automatically get the env from daq-kube

jcpunk replied:
I think this comment goes in the other service....

TiagoTAlves committed:
added RUN_START to the right place

TiagoTAlves committed:
added a foreignkey between RunReg tables

TiagoTAlves committed:
asthetic changes

alessandrothea commented:
Hi, thanks for updating the PR description @TiagoTAlves .

One comment about the high-level goals (I haven't had a chance to look at the code yet).

Simplifying the codebase and switching to `SQLAlchemy` is great, but I'm a little concerned by the following line

> All mentions of Oracle have been replaced with PostgresSQL.

It is a requirement for these services to work with both Oracle (CERN) and PostgresSQL (FNAL, Pocket).
I see that `SQLAlchemy` can support both backends, so is this what you meant with that sentence above?
If so, can the Oracle mode be tested at CERN before moving forward? 
Alessandro

jcpunk commented on runnumber-rest/entrypoint.sh:
```diff
@@ -1,15 +1,8 @@
 
-ensure_required_variables "RNURI RNUSER RNPASS"
-
-export LD_LIBRARY_PATH=/usr/lib/oracle/12.1/client64/lib/:$LD_LIBRARY_PATH
```
If retaining Oracle compatibility is required, this `LD_LIBRARY_PATH` bit will probably still be needed in addition to whatever package provides the client libraries.

TiagoTAlves committed:
testing for self-doc

TiagoTAlves committed:
database changes

TiagoTAlves committed:
changes to api and rest to make it work

TiagoTAlves committed:
asthetic changes

TiagoTAlves committed:
removed irrelevant stuff from merging

TiagoTAlves committed:
more aesthetic changes

jcpunk commented on runnumber-rest/rest.py:
```diff
@@ -10,12 +10,13 @@
 from api import app, db
-
-# setenv DEBUG=True to enable debug mode
-DEBUG = os.getenv("DEBUG", "False").lower() == "true"
```
Why revert this?

TiagoTAlves replied:
This was a mistake when merging Pierre's PR

TiagoTAlves replied:
Will change it back

jcpunk commented on runregistry-rest/rest.py:
```diff
@@ -1,11 +1,11 @@
 #!/usr/bin/env python3
-"""
+'''
```
Any reason to switch these?

TiagoTAlves replied:
Also remenants of the merge with Pierres PR

TiagoTAlves committed:
bad changes made reverted

TiagoTAlves committed:
merging with develop

jcpunk commented on runnumber-rest/rest.py:
```diff
@@ -1,4 +1,25 @@
+<<<<<<< HEAD
```
Looks like the merge failed and left behind artifacts

TiagoTAlves replied:
This was a simple forgot to save changes before committing

TiagoTAlves committed:
Merge failed, fixing it

jcpunk commented:
Since we're looking to have a devel instance running along side the production instance for testing.  Perhaps an env var or something to add a `development_` prefix to the table names might be useful.  That way we can test the same database as production, but without interference?

TiagoTAlves committed:
Merge branch 'develop' into riehecky/runnumber_reorg

jcpunk commented on runregistry-rest/entrypoint.sh:
```diff
@@ -1,14 +1,14 @@
+ensure_required_variables "DATABASE_URI"
 
-export LD_LIBRARY_PATH=/usr/lib/oracle/12.1/client64/lib/:$LD_LIBRARY_PATH
+if [[ ! -d /uploads ]]; then
```
I think this path needs to be fiddled with

TiagoTAlves replied:
What is the new path then?

jcpunk replied:
https://github.com/DUNE-DAQ/microservices/commit/15a1b5b5955c397e98c86577d0ffdabc9c316334  should be sufficient.  The program has already run `cd` at the dir test point.

TiagoTAlves committed:
added two env_variables, one to establish what database (Oracle or Postgres) and what deployment pod (prod or dev) that the microservices gets deployed to

TiagoTAlves commented:
> Since we're looking to have a devel instance running along side the production instance for testing. Perhaps an env var or something to add a `development_` prefix to the table names might be useful. That way we can test the same database as production, but without interference?

Would this be done in the microservices repo or the pocket repo?

TiagoTAlves committed:
added more env_variable conditions for oracle/postgresql and dev/prod

TiagoTAlves committed:
Merge branch 'develop' into riehecky/runnumber_reorg

TiagoTAlves committed:
made some changes to api's that I had forgotten to change when converting to sqlalchemy

jcpunk commented:
The microservices repo would need to have sane default logic, and then I'd setup a tweak for the kubernetes manifest to set the env var for devel/prod

TiagoTAlves commented:
> The microservices repo would need to have sane default logic, and then I'd setup a tweak for the kubernetes manifest to set the env var for devel/prod

The logic should still hold though because you only require to put prod/dev if and only if you are you set the environment variable as oracle, not postgresql

TiagoTAlves committed:
small corrections to previous commits

TiagoTAlves committed:
small changes for oracle test

TiagoTAlves committed:
changes in a hurry for oracle testing

jcpunk committed:
cleanup leaks from daq-kube

jcpunk committed:
python black/isort

jcpunk committed:
database type is derrived from database uri

jcpunk committed:
Switch to open source oracle client

fewer features, but easy to redistribute/build off site

TiagoTAlves committed:
oracles doesn't like quotation marks so changed to inverted commas

TiagoTAlves committed:
aesthetic changes to database.py in both runservices

jcpunk committed:
Fix exception catch

jcpunk committed:
Switch to fstrings

TiagoTAlves committed:
rn isn't used in run registry so changed it to run_number

TiagoTAlves committed:
changes to runregistry and runnumber to fix api queries

TiagoTAlves committed:
further changes to get runregistry working

TiagoTAlves committed:
foreign key and formatting

jcpunk committed:
Ignore accidentially compiled python objects

TiagoTAlves committed:
changes to where function

TiagoTAlves committed:
making previous queries from oracle output the same as with new image

jcpunk committed:
Try and unify/simplify service/sqlalchemy

jcpunk committed:
Also include path

jcpunk committed:
Add indexes and maybe compat url paths?

jcpunk committed:
Setup compat paths for old and new api

jcpunk committed:
Run through black

jcpunk committed:
Fix typo assignment

jcpunk committed:
Run isort

jcpunk committed:
Minor spaces

jcpunk committed:
Asked Chat GPT to simplify the codebase

jcpunk committed:
Need to import after `db` is init

jcpunk committed:
Simplify a bit more

jcpunk committed:
Since we're using the default names, just let the names default

jcpunk committed:
Try to make database start at the right run number

jcpunk committed:
Fix typo in debug messages

jcpunk committed:
template out the help page. Not sure if this is cleaner or not...

jcpunk committed:
Not nested enough

jcpunk committed:
Update doc

jcpunk committed:
Use positive logic

jcpunk committed:
Drop default lines

jcpunk committed:
Oracle thick client not required

TiagoTAlves committed:
make response belongs to flask

TiagoTAlves committed:
changed names of directories (so can be called for importing) and deleted database and authentication from runnumber_rest and called the same python files from runregistry_rest

TiagoTAlves committed:
changed runnumber api to macth runregistry api

TiagoTAlves committed:
changed the comments and returns to match oracle

TiagoTAlves committed:
some returns had to many nested []

jcpunk committed:
Revert "template out the help page. Not sure if this is cleaner or not..."

This reverts commit ded3fc235e9bf60e2169a0921ca37daf7c1f7d99.

TiagoTAlves committed:
Merge pull request #90 from DUNE-DAQ/titavare/runservices_reorg

Making some changes on a different branch, just in case you're unhappy with them and want to do them in a different way

TiagoTAlves committed:
changed all modules so they're all calling from microservices as the python path

TiagoTAlves committed:
added pythonpath to dockerfile for module finding

TiagoTAlves committed:
changing app route / in api

jcpunk committed:
Try to make the runregistry application fully replace runnumber.

It should now be able to listen on both ports and evaluate queries
within a sql transaction set for better atomic bits.

jcpunk committed:
Fix path

jcpunk committed:
Drop default lines

TiagoTAlves committed:
combining some changes made in other PR to make it work in both postgresql and oracle

jcpunk commented:
Why did you close this out?

TiagoTAlves commented:
I closed it because people were getting confused between both pull requests, once the other was merged I'd reopen the request. If you think this is the wrong approach we could open it again

jcpunk commented:
This is the one with the actual python changes.  If you're talking about the daq-kube pr, I'd close that one and re-open this instead.
